src/slice/quicksort.rs

```diff
@@ -5,16 +5,33 @@
 //! `rayon_core::join`.
 
 use std::cmp;
+use std::marker::PhantomData;
 use std::mem::{self, MaybeUninit};
 use std::ptr;
 
 /// When dropped, copies from `src` into `dest`.
-struct CopyOnDrop<T> {
+struct CopyOnDrop<'a, T> {
     src: *const T,
```

Comment:
Looking at where we create `CopyOnDrop`, I think this could just be `&'a T`?
(But the `marker` is still useful for invariance in `T`.)

Or I guess this is where you were suggesting unifying `ManuallyDrop`... TBH it feels a little uncomfortable that we're even creating a live `T` copy for `ManuallyDrop::new` in the first place. We could avoid *that* with `src: MaybeUninit<T>`, initialized by copy from a source pointer. The lifetime would just become `'dest`, turbofish-constrained as `'v` when we create it. What do you think?

Reply:
Yeah I think it's definitely _cleaner_ to unify it with ManuallyDrop but it's not safer or anything. I don't want to do the broader refactor here though, I was trying to get a small defense in depth win.